### PR TITLE
Download MNIST dataset from remote URL

### PR DESCRIPTION
**Pull request by spencerkohan**
This PR is intended to fix the second item listed in [this issue](https://github.com/tensorflow/swift-models/issues/206):

> it currently relies on hardcoded paths, which will break if you take the dataset outside of that project and try to use it in something like Colab.

In this PR, the datasets are instead downloaded from a remote URL, so they are not fragile with respect to the current working directory where the MNIST dataset is instantiated from.

The implementation in this PR brings the MINST dataset in line with how the CFAR10 dataset is loaded.

### CODE REVIEW

File: Datasets/DatasetUtils.swift [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+        do {
+            print("Loading local data at: \(localURL.path)")
+            let data = try Data(contentsOf: localURL)
+            print("Succesfully loaded resource: \(filename)")
+            return data
+        } catch {
+            fatalError("Failed to contents of resource: \(localURL)")
+        }
+
```
Comment (BradLarson) [inline Datasets/DatasetUtils.swift]: Minor style thing, but I don't think we need the extra spaces inserted after many of the closing (and some opening) braces here.

File: Datasets/DatasetUtils.swift [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    }
+
+    struct ResourceDefinition {
+
+        let filename: String
+        let remoteRoot: URL
+        let localStorageDirectory: URL
+
+        var localURL: URL {
+            return localStorageDirectory.appendingPathComponent(filename)
```
Comment (rxwei) [inline Datasets/DatasetUtils.swift]: Omit `return` for consistency with other single-expression computed properties defined in this file.
```suggestion
            localStorageDirectory.appendingPathComponent(filename)
```

File: Datasets/DatasetUtils.swift [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+            print("Succesfully loaded resource: \(filename)")
+            return data
+        } catch {
+            fatalError("Failed to contents of resource: \(localURL)")
+        }
+
+    }
+
+    struct ResourceDefinition {
+
```
Comment (rxwei) [inline Datasets/DatasetUtils.swift]: Remove empty line. In this code base we tend not to add leading or trailing empty lines in type declarations.

File: Datasets/DatasetUtils.swift [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+        print("Loading resource: \(filename)")
+
+        let resource = ResourceDefinition(
+            filename: filename,
+            remoteRoot: remoteRoot,
+            localStorageDirectory: localStorageDirectory)
+
+        let localURL = resource.localURL
+
+        // Fetch from remote if the file is not available locally
```
Comment (rxwei) [inline Datasets/DatasetUtils.swift]: * Use three slashes for doc comments.
* Start a doc comment with a third-person singular verb.
* End a doc comment with a period.

File: Datasets/DatasetUtils.swift [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+// limitations under the License.
+
+import Foundation
+
+#if canImport(FoundationNetworking)
+    import FoundationNetworking
+#endif
+
+public struct DatasetUtils {
+
```
Comment (rxwei) [inline Datasets/DatasetUtils.swift]: Remove empty line.

File: Datasets/DatasetUtils.swift [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+        let task = Process()
+        task.executableURL = URL(fileURLWithPath: gunzipLocation)
+        task.arguments = [archivePath]
+        do {
+            try task.run()
+            task.waitUntilExit()
+        } catch {
+            fatalError("Failed to extract \(archivePath) with error: \(error)")
+        }
+
```
Comment (rxwei) [inline Datasets/DatasetUtils.swift]:
```suggestion
```

File: Datasets/DatasetUtils.swift [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+    }
+
```
Comment (rxwei) [inline Datasets/DatasetUtils.swift]:
```suggestion
```

File: Datasets/DatasetUtils.swift [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+        } catch {
+            fatalError("Failed to fetch and save resource with error: \(error)")
+        }
+        print("Archive saved to: \(archiveLocation.path)")
+
+        extractArchive(for: resource)
+    }
+
+    static func extractArchive(for resource: ResourceDefinition) {
+
```
Comment (rxwei) [inline Datasets/DatasetUtils.swift]:
```suggestion
```

File: Datasets/MNIST/MNIST.swift [hunk 9]
```diff
@@ -61,36 +67,32 @@ extension Tensor {
...
-        let data = try! Data(contentsOf: filePath, options: [])
-        return [UInt8](data)
+fileprivate func fetchDataset(
+    localStorageDirectory: URL,
+    imagesFilename: String,
+    labelsFilename: String,
+    flattening: Bool,
+    normalizing: Bool
+) -> (images: Tensor<Float>, labels: Tensor<Int32>) {
+
```
Comment (rxwei) [inline Datasets/MNIST/MNIST.swift]:
```suggestion
```

File: Datasets/DatasetUtils.swift [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+    public static let curentWorkingDirectoryURL = URL(
+        fileURLWithPath: FileManager.default.currentDirectoryPath)
+
+    public static func fetchResource(
+        filename: String,
+        remoteRoot: URL,
+        localStorageDirectory: URL = curentWorkingDirectoryURL
+    ) -> Data {
+
```
Comment (rxwei) [inline Datasets/DatasetUtils.swift]:
```suggestion
```

File: Datasets/DatasetUtils.swift [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+        }
+
+        var remoteURL: URL {
+            remoteRoot.appendingPathComponent(filename).appendingPathExtension("gz")
+        }
+
+        var archiveURL: URL {
+            localURL.appendingPathExtension("gz")
+        }
+
```
Comment (rxwei) [inline Datasets/DatasetUtils.swift]:
```suggestion
```

File: Datasets/DatasetUtils.swift [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import Foundation
+
+#if canImport(FoundationNetworking)
+    import FoundationNetworking
+#endif
+
+public struct DatasetUtils {
```
Comment (rxwei) [inline Datasets/DatasetUtils.swift]: Avoid abbreviations.
```suggestion
public struct DatasetUtilities {
```

File: Datasets/MNIST/MNIST.swift [hunk 13]
```diff
@@ -101,9 +103,11 @@ func readMNIST(imagesFile: String, labelsFile: String, flattening: Bool, normali
         return (images: flattenedImages, labels: Tensor(labels))
     } else {
         return (
-            images: Tensor(shape: [rowCount, 1, imageHeight, imageWidth], scalars: images)
+            images:
+                Tensor(shape: [rowCount, 1, imageHeight, imageWidth], scalars: images)
                 .transposed(withPermutations: [0, 2, 3, 1]) / 255,  // NHWC
```
Comment (rxwei) [inline Datasets/MNIST/MNIST.swift]:
```suggestion
                     .transposed(withPermutations: [0, 2, 3, 1]) / 255,  // NHWC
```

File: Datasets/MNIST/MNIST.swift [hunk 14]
```diff
@@ -101,9 +103,11 @@ func readMNIST(imagesFile: String, labelsFile: String, flattening: Bool, normali
...
             labels: Tensor(labels)
         )
     }
+
```
Comment (rxwei) [inline Datasets/MNIST/MNIST.swift]:
```suggestion
```

File: Datasets/DatasetUtils.swift (hunk 4 above)
Comment (spencerkohan) [reply Datasets/DatasetUtils.swift]: This comment seemed redundant with the print statement, and it's pretty self-explanatory from the code anyway, so I just removed it.  Let me know if you think it's better to have a doc comment there.

**Comment by spencerkohan** [comment]
@BradLarson @rxwei I believe I have made all the requested changes.  Thanks for taking the time to review.

File: Datasets/DatasetUtilities.swift [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+        var remoteURL: URL {
+            remoteRoot.appendingPathComponent(filename).appendingPathExtension("gz")
+        }
+
+        var archiveURL: URL {
+            localURL.appendingPathExtension("gz")
+        }
+    }
+
+    static func fetchFromRemoteAndSave(resource: ResourceDefinition) {
```
Comment (rxwei) [inline Datasets/DatasetUtilities.swift]: The `resource:` argument label is seemingly only repeating the type information `ResourceDefinition`, and it doesn't affect the grammaticality of the phrase at call sites. I'd suggest dropping the argument label.
```suggestion
    static func fetchFromRemoteAndSave(_ resource: ResourceDefinition) {
```

**Pull request merged**